lib/nostrum/shard/session.ex

```diff
@@ -86,63 +111,63 @@ defmodule Nostrum.Shard.Session do
 
     case from_handle do
       {new_state, reply} ->
-        {:reply, {:binary, reply}, new_state}
+        :ok = :gun.ws_send(state.conn, {:binary, reply})
+        {:noreply, new_state}
 
       new_state ->
-        {:ok, new_state}
+        {:noreply, new_state}
     end
   end
 
+  def handle_info({:gun_ws, _conn, _stream, {:close, errno, reason}}, state) do
+    Logger.info("Shard websocket closed (errno #{errno}, reason #{inspect(reason)}).")
+    {:noreply, state}
+  end
+
+  def handle_info(
+        {:gun_down, _conn, _proto, _reason, _killed_streams, _unprocessed_streams},
+        state
+      ) do
+    {:noreply, state}
+  end
+
+  def handle_info({:gun_up, worker, _proto}, state) do
+    stream = :gun.ws_upgrade(worker, @gateway_qs)
+    await_ws_upgraded(worker, stream)
+    Logger.warn("Reconnected after connection broke.")
+    {:noreply, state}
+  end
+
   def handle_cast({:status_update, payload}, state) do
-    {:reply, {:binary, payload}, state}
+    :ok = :gun.ws_send(state.conn, {:binary, payload})
+    {:noreply, state}
   end
 
   def handle_cast({:update_voice_state, payload}, state) do
     {:reply, {:binary, payload}, state}
   end
 
   def handle_cast({:request_guild_members, payload}, state) do
-    {:reply, {:binary, payload}, state}
+    :ok = :gun.ws_send(state.conn, {:binary, payload})
+    {:noreply, state}
   end
 
   def handle_cast(:heartbeat, %{heartbeat_ack: false} = state) do
     Logger.warn("heartbeat_ack not received in time, disconnecting")
-    {:close, state}
+    :ok = :gun.close(state.conn)
+    {:stop, {:heartbeat_ack_timeout, state.shard_num}, state}
```

Comment:
I'm not sure we should stop the process here because then we can't resume through discord. We should just cancel the timer ref here and let it restart gracefully.

Reply:
But that's what I'm doing, huh?

I think that if we receive this we might not have a timer.

Reply:
I mean, in a way we are cancelling the timer ref because we're stopping the gen_server process. I was trying to say that we should not stop the process and instead just cancel the timer ref and let gun try to reconnect for us, and in doing so start another heartbeat.

Reply:
:shipit: